File: app/src/main/java/com/afoxxvi/alopex/component/notify/NotifyGroup.java
package com.afoxxvi.alopex.component.notify;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;

import androidx.databinding.BaseObservable;
import androidx.databinding.Bindable;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

public class NotifyGroup extends BaseObservable {
    private final String packageName;
    private String appName;
    private Drawable appIcon;
    private final int packageId;
    private final List<Notify> notifyList;

    public NotifyGroup(Context context, String packageName, int packageId) {
        this.packageName = packageName;
        this.packageId = packageId;
        this.notifyList = new ArrayList<>();
        PackageManager packageManager = context.getPackageManager();
        try {
            PackageInfo info = packageManager.getPackageInfo(packageName, PackageManager.GET_ACTIVITIES | PackageManager.GET_META_DATA);
            this.appName = info.applicationInfo.loadLabel(packageManager).toString();
            this.appIcon = info.applicationInfo.loadIcon(packageManager);
        } catch (PackageManager.NameNotFoundException e) {
            this.appName = null;
            this.appIcon = null;
        }
    }

    @Bindable
    public String getPackageName() {
        return packageName;
    }

    @Bindable
    public String getAppName() {
        return appName == null ? getPackageName() : appName;
    }

    @Bindable
    public Drawable getAppIcon() {
        return appIcon;
    }

    public int getPackageId() {
        return packageId;
    }

    @Bindable
    public String getLatestTitle() {
        if (!notifyList.isEmpty()) {
            return notifyList.get(0).getTitle();
        }
        return "<Empty>";
    }

    @Bindable
    public String getLatestText() {
        if (!notifyList.isEmpty()) {
            return notifyList.get(0).getText();
        }
        return "<No notification recently>";
    }

    @Bindable
    public String getLatestTimeText() {
        if (!notifyList.isEmpty()) {
            return notifyList.get(0).getDateText();
        }
        return "";
    }

    @Bindable
    public Integer getNotifyCount() {
        return notifyList.size();
    }

    public LocalDateTime getLatestTime() {
        if (!notifyList.isEmpty()) {
            return notifyList.get(0).getTime();
        }
        return LocalDateTime.MIN;
    }

    public LocalDateTime getEarliestNotifyTime() {
        if (!notifyList.isEmpty()) {
            return notifyList.get(notifyList.size() - 1).getTime();
        }
        return LocalDateTime.now();
    }

    public List<Notify> getNotifyList() {
        return notifyList;
    }
}

File: app/src/main/java/com/afoxxvi/alopex/component/notify/NotifyGroup.kt
package com.afoxxvi.alopex.component.notify

import android.content.Context
import android.content.pm.PackageManager
import android.content.pm.PackageManager.NameNotFoundException
import android.graphics.drawable.Drawable
import androidx.databinding.BaseObservable
import androidx.databinding.Bindable
import java.time.LocalDateTime

class NotifyGroup(context: Context, @get:Bindable val packageName: String, val packageId: Int) : BaseObservable() {
    private var appName: String? = null

    @get:Bindable
    var appIcon: Drawable? = null
    private val notifyList: MutableList<Notify>

    init {
        notifyList = ArrayList()
        val packageManager = context.packageManager
        try {
            val info = packageManager.getPackageInfo(packageName, PackageManager.GET_ACTIVITIES or PackageManager.GET_META_DATA)
            appName = info.applicationInfo.loadLabel(packageManager).toString()
            appIcon = info.applicationInfo.loadIcon(packageManager)
        } catch (e: NameNotFoundException) {
            appName = null
            appIcon = null
        }
    }

    @Bindable
    fun getAppName(): String {
        return appName ?: packageName
    }

    @get:Bindable
    val latestTitle: String?
        get() = if (notifyList.isNotEmpty()) {
            notifyList[0].title
        } else "<Empty>"

    @get:Bindable
    val latestText: String?
        get() = if (notifyList.isNotEmpty()) {
            notifyList[0].text
        } else "<No notification recently>"

    @get:Bindable
    val latestTimeText: String
        get() = if (notifyList.isNotEmpty()) {
            notifyList[0].dateText
        } else ""

    @get:Bindable
    val notifyCount: Int
        get() = notifyList.size
    val latestTime: LocalDateTime
        get() = if (notifyList.isNotEmpty()) {
            notifyList[0].time
        } else LocalDateTime.MIN
    val earliestNotifyTime: LocalDateTime
        get() = if (notifyList.isNotEmpty()) {
            notifyList[notifyList.size - 1].time
        } else LocalDateTime.now()

    private fun checkNotify(notify: Notify) {}
    fun addNotify(notify: Notify) {
        notifyList.add(notify)
    }

    fun addNotify(index: Int, notify: Notify) {
        notifyList.add(index, notify)
    }

    fun getNotify(index: Int): Notify {
        return notifyList[index]
    }

    fun limitCount(max: Int) {
        notifyList.subList(max, notifyCount).clear()
    }
}